design/FY2018/NFP-Table-Refactor.md

```diff
@@ -0,0 +1,819 @@
+
+$$\begin{array}{rl}
+f^\prime(x_0) &= ( f(x_1) - f(x_{-1}) ) / ( x_1 - x_{-1}) \\
+f^\prime(x_1) &= ( f(x_2) - f(x_0) ) / ( x_2 - x_0)
+\end{array}$$
+
+When the hypercube is at the edge of the grid, Catmull-Rom simply extends the slope of the final segment for defining the slope terms: i.e., if there is no $x_{-1}$, we substitute $x_0$ into the $f^\prime(x_0)$ formula.
+
+If a mix of interpolation methods are specified among the included dimensions, the library will perform that mix as requested.
+
+If the lookup point is beyond the grid edge on any axis, the library will perform the requested extrapolation method (linear or constant) on that dimension while proceeding with interpolation along any in-bounds dimension. If a "do-not-extrapolate-beyond" boundary is specified, the library will perform a constant extrapolation from that boundary outward (in that dimension), and it will return the resulting numerical answer along with a warning that the request was outside the boundaries.
+
+
+## Source Code Design
+
+The main component of the source code design is the addition of a new tabular data interpolation library (currently called "Btwxt" [betwixt]). The purpose of this library is to isolate the interpolation computation and data structures from the rest of EnergyPlus in a fashion that is easily unit tested and can be easily ported to other C++ programs.
+
+The other component is the integration of Btwxt in EnergyPlus. This will involve a modest refactoring of how EnergyPlus handles table data structures.
+
+### Btwxt
+
+Btwxt is a general-purpose N-dimensional interpolation library for gridded data. After reviewing existing open-source libraries, we believe it is sensible to write a new library.
+
+The library will accept:
+*   a collection of N vectors representing the input variables,
+*   attributes of each of the input vectors describing preferred interpolation and extrapolation methods, and
+*   an array of (or collection of arrays of) values that map onto the grid defined by the input vectors.
+
+Together, the input vectors and array(s) of output values will define a performance space. The library will allow querying of a defined performance space for a well-defined point within the grid.
+
+*   A point fully inside the grid boundaries will receive an interpolated value.
+    *   The library will support (a) linear and (b) cubic interpolation.
+    *   The library structure will allow for alternative interpolation methods to be added later.
+*   A point outside the grid boundaries will receive an extrapolated value.
+    *   The library will support (a) linear and (b) constant extrapolation
+    *   The library will allow each input dimension to define its extrapolation method and an extrapolation limit.
+    *   Each input dimension will have the option of specifying outer boundaries beyond which the library will not interpolate.
+    *   Each input dimension will have the option of disallowing any extrapolation.
```

Comment:
@nealkruis It'd be super helpful for me it you included a snippet of code that shows what use looks like in the context of E+ (e.g. taking two IVs and getting one value). Maybe this section is the wrong place for that, but it could also fit in the E+ integration section.

Reply:
@jasondegraw [here is the function](https://github.com/NREL/EnergyPlus/pull/6749/files#diff-6bab88f88cad01231c7611b162099cdaR5400) that is called for the new tables. The main call into Btwxt is [near the bottom](https://github.com/NREL/EnergyPlus/pull/6749/files#diff-6bab88f88cad01231c7611b162099cdaR5441) of that function. The IVs are passed in as a std::vector, and the interpolator returns a std::vector of results corresponding to the values of each table that share the same parameter space (e.g., EIR and Capacity as f(T)). Maybe a call would work well to walk you through the process?